test/test-prepare-check.c

```diff
@@ -0,0 +1,138 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <event2/event.h>
+#include <event2/event_watcher.h>
+#include <assert.h>
+
+static int iteration = 0;
+static int prepare_callback_1_count = 0;
+static int prepare_callback_2_count = 0;
+static int check_callback_1_count = 0;
+static int check_callback_2_count = 0;
+static struct timeval start_time = { 0, 0 };
+static struct timeval end_time = { 0, 0 };
+
+#define log_callback \
+	printf("%s:\n", __func__); \
+	printf("  now: %ld.%06ld\n", info->now.tv_sec, info->now.tv_usec); \
+	printf("  timeout: %ld.%06ld\n", info->timeout->tv_sec, info->timeout->tv_usec);
+
+void
+prepare_callback_1(struct event_watcher *watcher, const struct event_watcher_cb_info *info)
+{
+	log_callback;
+	++prepare_callback_1_count;
+
+	/* prepare_callback_1 should always fire before prepare_callback_2, and before both check callbacks */
+	assert(prepare_callback_1_count > prepare_callback_2_count);
+	assert(prepare_callback_1_count > check_callback_1_count);
+	assert(prepare_callback_1_count > check_callback_2_count);
+
+	/* if we've just scheduled the timeout event at the beginning of the iteration, save the current time and assert that the timeout is what we set */
+	if (start_time.tv_sec == 0) {
+		start_time = info->now;
+		assert(info->timeout->tv_sec == 1);
+		assert(info->timeout->tv_usec == 0);
+	}
+}
+
+void
+prepare_callback_2(struct event_watcher *watcher, const struct event_watcher_cb_info *info)
+{
+	log_callback;
+	++prepare_callback_2_count;
+
+	/* prepare_callback_2 should only fire on the first iteration, and should fire before both check callbacks */
+	assert(iteration == 0);
+	assert(prepare_callback_2_count > check_callback_1_count);
+	assert(prepare_callback_2_count > check_callback_2_count);
+}
+
+void
+check_callback_1(struct event_watcher *watcher, const struct event_watcher_cb_info *info)
+{
+	log_callback;
+	++check_callback_1_count;
+
+	/* check_callback_1 should always fire before check_callback_2 */
+	assert(check_callback_1_count > check_callback_2_count);
+
+	/* save the end time, in case the timeout fires this time through the event loop */
+	end_time = info->now;
+}
+
+void
+check_callback_2(struct event_watcher *watcher, const struct event_watcher_cb_info *info)
+{
+	log_callback;
+	++check_callback_2_count;
+
+	/* check_callback_2 should only fire on the first iteration */
+	assert(iteration == 0);
+}
+
+void
+timeout_callback(evutil_socket_t fd, short events, void *arg)
+{
+	printf("timeout_callback\n");
+
+	/* the duration between the start and end times should be at least 1 second */
+	assert(end_time.tv_sec >= start_time.tv_sec + 1);
+}
+
+int
+main(int argc, char **argv)
+{
+	struct event_base *base = event_base_new();
+
+	/* install prepare and check watchers */
+	event_watcher_prepare_new(base, &prepare_callback_1);
```

Comment:
These tests catch the regression you spotted around check/dispatch ordering, I'm guessing, can you call that out in a comment somewhere?

Reply:
Yeah, I should write a summary at the top... there's a fair bit going on here.